Retrieve blocks from multiple nodes and add a test-case when bad blocks are retrieved

AndrejMitrovic opened the pull request.

AndrejMitrovic committed:
Short-circuit when blocks fail validation.

AndrejMitrovic committed:
Add test-case when a node sends bad blocks

The receiving node should reject the blocks.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -351,6 +354,17 @@ public final class TestNode : Node, TestAPI
     }
+
+    /// Add a set of invalid blocks to the Ledger, skipping validation
+    public override void makeBadBlocks (size_t height, size_t count)
```
This just makes empty blocks right ? Not "meaningfully bad" blocks.

Geod24 commented:
I'm not sure about the dependency injection on the `Ledger` level. It does not sound like something that should be replaced.

AndrejMitrovic commented:
Actually come to think of it, I could make a fake node that returns invalid data directly in the `getBlocksFrom` API, without touching the Ledger at all.

Geod24 commented:
Commit 1 LVGTM.
Commit 2 I'm skeptical because just generating empty blocks does not sound generic enough to be the default implementation. I'd much rather have a `Block` being sent from the client. I *think* we could make `makeBlock` be `pure` (so that we get `immutable` blocks).
Commit 3 is not used.
Commit 4 LGTM

AndrejMitrovic commented:
3 is technically used by 4, but I will rewrite this to use a customly inherited Node. It's better than dependency injection.

AndrejMitrovic committed:
Attempt retrieving blocks from multiple nodes

Previously, the node retrieved blocks from the node
which had the highest height.

However this opens the possibility for a DoS,
as a node can simply lie about having the highest
height.

Instead, try downloading from all nodes until a
valid (!) block with the highest block height is
retreived from the network.

AndrejMitrovic commented:
Reworked the PR.

Instead of injecting dependencies, I made a custom `TestNode` subclass which returns invalid blocks.

Note that currently there is limited block validation (only the block height is checked). This will be improved in future PRs.

AndrejMitrovic committed:
Make two fields / methods public in tests

It's required to be able to access the 'tbn'
from TestAPIManager subclasses and call
createNewNode() from makeTestNetwork().

AndrejMitrovic commented:
Rebased.

Geod24 commented on source/agora/network/NetworkManager.d:
```diff
@@ -235,35 +241,34 @@ public class NetworkManager
 
-            logInfo("Retrieving latest blocks..");
-            auto node = node_pair[1];
+        LNextNode: foreach (pair; node_pairs.retro) try
```
You could just change the ordering and wouldn't need `retro`, no ?

AndrejMitrovic replied:
LOL

Geod24 commented on source/agora/test/NetworkManager.d:
```diff
@@ -49,3 +49,114 @@ unittest
+        /// Initialize a new node
+        public override void createNewNode (PublicKey address, Config conf)
+        {
+            static bool first_node = true;
```
`first_node == (this.nodes.length == 0)`

AndrejMitrovic replied:
It's a static initializer, that can't work.

AndrejMitrovic replied:
Ah nevermind.

AndrejMitrovic replied:
See what I mean? My brain..

Geod24 commented on source/agora/test/NetworkManager.d:
```diff
@@ -49,3 +49,114 @@ unittest
+    auto txes = makeChainedTransactions(getGenesisKeyPair(), null, 10);
+    txes.each!(tx => node_good.putTransaction(tx));
+
+    // at this point both the good node and bad node have same amount of blocks
```
```suggestion
    // at this point both the good node and bad node have same amount of blocks, but bad node pretends to have + 10
```

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -131,7 +131,7 @@ public class TestAPIManager
 
     /// Initialize a new node
-    protected void createNewNode (PublicKey address, Config conf)
+    public void createNewNode (PublicKey address, Config conf)
```
It seems `protected` is enough, according to the next commit

AndrejMitrovic replied:
`createNewNode` needs to be public as it's accessed from the `Base` module, and DMD has a bug: https://issues.dlang.org/show_bug.cgi?id=20002

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -245,7 +245,7 @@ public class TestNetworkManager : NetworkManager
 
     /// Workaround compiler bug that triggers in `std.concurrency`
-    protected __gshared std.concurrency.Tid[string] tbn;
+    public __gshared std.concurrency.Tid[string] tbn;
```
Same thing about `protected`

AndrejMitrovic replied:
Well `BadAPIManager` is not a subclass of `TestNetworkManager`, but of `TestAPIManager`, so it needs to be public. 

This is after I split the two classes into `TetAPIManager` and `TestNetworkManager`. I didn't feel it made sense to have it one class just in order to keep one `__gshared` field `protected.`

Geod24 replied:
Fair enough

AndrejMitrovic commented:
Fixed the rest of the bugs.
